# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [5]:

import warnings
warnings.simplefilter('ignore')
import pandas as pd
import numpy as np
import sqlite3


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [6]:
db = sqlite3.connect("Databases/GreatOutdoors/go_sales.sqlite")

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [7]:
product = pd.read_sql_query("SELECT * FROM product;", db)
country = pd.read_sql_query("SELECT * FROM country;", db)
sales_branch = pd.read_sql_query("SELECT * FROM sales_branch;", db)
retailer_site = pd.read_sql_query("SELECT * FROM retailer_site;", db)
product_type = pd.read_sql_query("SELECT * FROM product_type;", db)
returns = pd.read_sql_query("SELECT * FROM returned_item;", db)
order_details = pd.read_sql_query("SELECT * FROM order_details;", db)
sales_staff = pd.read_sql_query("SELECT * FROM sales_staff;", db)
order_header = pd.read_sql_query("SELECT * FROM order_header;", db)
returned_item  = pd.read_sql_query("SELECT * FROM returned_item;", db)
return_reason  = pd.read_sql_query("SELECT * FROM return_reason;", db)
target = pd.read_sql_query("SELECT * FROM SALES_TARGETData;", db)






Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [24]:
# product
# product['PRODUCTION_COST'].astype(np.double)
product['PRODUCTION_COST'] = product["PRODUCTION_COST"].astype("double")

filtered_production_price = (product['PRODUCTION_COST'] > 50) & (product['PRODUCTION_COST'] < 100)

product.loc[(filtered_production_price), ['PRODUCT_NAME', 'PRODUCTION_COST']]


PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [47]:
# product['PRODUCTION_COST'] = product["PRODUCTION_COST"].astype("double")
product['MARGIN'] = product["MARGIN"].astype("float")

marge_filter = (product['MARGIN'] < .2) | (product['MARGIN'] > .6)
product.loc[(marge_filter), :]

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE  PRODUCTION_COST  \
29              21         15-2-2011                 3            29.17   
32              24          5-3-2013                 4            62.50   
100             86         15-2-2011                15             2.00   
101             87         15-2-2011                15             2.00   
102             88         15-2-2011                15             2.00   
103             89         15-2-2011                15             2.50   
105             90         15-2-2011                15             2.50   

     MARGIN   PRODUCT_IMAGE LANGUAGE                  PRODUCT_NAME  \
29     0.17   P21CE1SB3.jpg       EN                Hibernator Pad   
32     0.17   P24CE1PK4.jpg       EN  Canyon Mule Climber Backpack   
100    0.67  P86OP4IR15.jpg       EN             BugShield Natural   
101    0.67  P86OP4IR15.jpg       EN               BugShield Spray   
102    0.70  P86OP4IR15.jpg       EN         BugShield Lotion Lite   
103    0.63  P86OP4IR15.jpg       EN              BugShield Lotion   
105    0.63  P86OP4IR15.jpg       EN             BugShield Extreme   

                                           DESCRIPTION TRIAL888  
29   Warms up quickly for snow camping comfort. The...        T  
32   This pack is perfect for day trips and short h...        T  
100  Contains no chemicals. Repels mosquitoes, blac...        T  
101  Repels black flies, ticks, mosquitoes, chigger...        T  
102  Repels black flies, ticks, mosquitoes, chigger...        T  
103  Repels black flies, ticks, mosquitoes, chigger...        T  
105  Highest amount of the active ingredient, 38%, ...        T

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [55]:
francs_countries = country['CURRENCY_NAME'] == 'francs'
country.loc[(francs_countries), :].sort_values(by=['COUNTRY'])

COUNTRY_CODE      COUNTRY LANGUAGE CURRENCY_NAME TRIAL879
7            17      Belgium       EN        francs        T
0             1       France       EN        francs        T
19            8  Switzerland       EN        francs        T

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [62]:
product['MARGIN'] = product["MARGIN"].astype("float")

marge_filter = product['MARGIN'] > .5
product.loc[(marge_filter), ["INTRODUCTION_DATE"]].drop_duplicates()

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [68]:
sales_branch

sales_branch.loc[:, ["ADDRESS1", "ADDRESS2"]]

ADDRESS1           ADDRESS2
0                     Schwabentor 35               None
1                   Leopoldstraße 36               None
2                Isafjordsgatan 30 C               None
3       7800, 756 - 6th Avenue. S.W.               None
4                   789 Yonge Street               None
5             1288 Dorchester Avenue               None
6                    299 Yale Avenue               None
7         1288 South Barrington Ave.               None
8                     10032 NW 186th               None
9                6c, rue de l'Église               None
10  Prol. Paseo de la Reforma No. 51  Lomas de Santa Fe
11               202-2-3 Hyakunincho        Shinjuku-ku
12                     543-225 Asahi           Asahi-ku
13                  2315 Queen's Ave            Level 2
14     Plaza de la Constitución, s/n               None
15             Avenida Paulista, 333   CJ 231 2o. Andar
16                     Kauppakatu 33               None
17             234-12, Kongdeok-Dong            Mapo-Ku
18      154, Choung-Cheng South Road               None
19          No. 1277 Beijing Xi Road               None
20                    Victoria House    25 Ebury Street
21                    3 Albany Court      Market Street
22                    Birchstraße 13               None
23                 Interleuvenlaan 2               None
24                 Jedleser Straße 7               None
25     75, rue du Faubourg St-Honoré               None
26                   Piazza Duomo, 1               None
27               Singelgravenplein 4      4e verdieping

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [75]:
dollars_countries = (country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar')
country.loc[(dollars_countries), ["COUNTRY"]].sort_values(by=['COUNTRY'])


COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [86]:
filter = (retailer_site['POSTAL_ZONE'].str[0] == "D") & (retailer_site['ADDRESS2'].notna())
retailer_site.loc[filter, ["ADDRESS1", "ADDRESS2", "CITY"]]

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [95]:
returns["RETURN_QUANTITY"] = returns["RETURN_QUANTITY"].astype("int64")
returns["RETURN_QUANTITY"].sum()

14306

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [116]:
# sales_branch["REGION"].notna()
not_empty_region = sales_branch["REGION"].notna()

sales_branch.loc[not_empty_region]["REGION"].count()

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [128]:
product['MARGIN'] = product['MARGIN'].astype("double")
laagste_marge = product['MARGIN'].min()
hoogste_marge = product['MARGIN'].max()
avg_marge = np.round(product['MARGIN'].agg(np.mean), 2)

frame = {
    "Laagste Marge": [laagste_marge],
    "Hoogste Marge": [hoogste_marge],
    "Avg Marge": [avg_marge],
}

pd.DataFrame(frame)

Laagste Marge  Hoogste Marge  Avg Marge
0           0.17            0.7        0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [133]:
empty_address2 = sales_branch["ADDRESS2"].isna()

sales_branch.loc[empty_address2]['SALES_BRANCH_CODE'].count()

19

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [149]:
avg_kostprijs = order_details["UNIT_SALE_PRICE"] < order_details["UNIT_PRICE"]


order_details["UNIT_PRICE"] = order_details["UNIT_PRICE"].astype("double")
order_details.loc[avg_kostprijs, "UNIT_PRICE"].agg(np.mean)

192.88697889934417

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [218]:
frame = pd.DataFrame()
roles = sales_staff["POSITION_EN"].drop_duplicates()

idx = 0
for role in roles:
    frame.at[idx,'role'] = role
    filter = sales_staff.loc[:, "POSITION_EN"].str.contains(role)
    frame.at[idx,'count'] = sales_staff.loc[filter, "POSITION_EN"].count()
    idx += 1
frame




role  count
0  Level 2 Sales Representative   39.0
1  Level 3 Sales Representative   24.0
2                Branch Manager   19.0
3               General Manager    1.0
4  Level 1 Sales Representative   13.0
5              Regional Manager    2.0
6              District Manager    4.0

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

IndexingError: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [75]:
merged_df = pd.merge(left=retailer_site, right=country, how='inner', left_on= 'COUNTRY_CODE', right_on='COUNTRY_CODE')

dutch_retailer_sites = merged_df["COUNTRY"] == "Netherlands"

merged_df.loc[dutch_retailer_sites, :]

RETAILER_SITE_CODE RETAILER_CODE                   ADDRESS1 ADDRESS2  \
164                272           141                   Beets 36     None   
165                273           141               Perestraat 1     None   
166                274           141  Laan van Meerdervoort 966     None   
167                275           141             Kalfjeslaan 47     None   
168                276           141               Startbaan 11     None   
169                277           142      Noorderspoorsingel 28     None   
170                278           143              Hogehilweg 13     None   
171                279           143             Spoorstraat 43     None   
173                280           144             Westhavenweg 8     None   
174                281           144              Westzeedijk 4     None   
175                282           144            Gooiergracht 91     None   
176                283           144            Hoofdstraat 150     None   
177                284           145                Olde Diek 2     None   
178                285           145     Amsterdamsestraatweg 8     None   
179                286           145        Krayenhoffstraat 11     None   
180                287           145         Overschiestraat 19     None   
181                288           145              Havenbuurt 12     None   
182                289           146               Duivendaal 1     None   
184                290           146              Voorstraat 37     None   
185                291           146         Stadhouderslaan 18     None   

             CITY REGION POSTAL_ZONE COUNTRY_CODE ACTIVE_INDICATOR TRIAL888  \
164         Beets   None     1475 JC            7                1        T   
165      Deventer   None     7412 BV            7                1        T   
166      Den Haag   None     2564 AV            7                1        T   
167    Amstelveen   None     1181 BN            7                1        T   
168    Amstelveen   None     1185 XP            7                1        T   
169     Groningen   None     9716 JE            7                1        T   
170     Amsterdam   None     1000 BA            7                1        T   
171    Varsseveld   None     7051 CH            7                1        T   
173     Amsterdam   None     1014 BD            7                1        T   
174        Muiden   None     1398 BB            7                1        T   
175         Laren   None     1251 VC            7                1        T   
176   Pieterburen   None     9968 AH            7                1        T   
177         Laren   None     7245 VS            7                1        T   
178      Maarssen   None     3604 AA            7                1        T   
179     Amsterdam   None     1018 DE            7                1        T   
180     Amsterdam   None     1062 XC            7                1        T   
181        Marken   None     1156 AL            7                1        T   
182    Wageningen   None     6701 AP            7                1        T   
184  Nieuweschans   None     9693 EE            7                1        T   
185     Rotterdam   None     3051 HG            7                1        T   

         COUNTRY LANGUAGE CURRENCY_NAME TRIAL879  
164  Netherlands       EN      guilders        T  
165  Netherlands       EN      guilders        T  
166  Netherlands       EN      guilders        T  
167  Netherlands       EN      guilders        T  
168  Netherlands       EN      guilders        T  
169  Netherlands       EN      guilders        T  
170  Netherlands       EN      guilders        T  
171  Netherlands       EN      guilders        T  
173  Netherlands       EN      guilders        T  
174  Netherlands       EN      guilders        T  
175  Netherlands       EN      guilders        T  
176  Netherlands       EN      guilders        T  
177  Netherlands       EN      guilders        T  
178  Netherlands       EN      gu

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [80]:
merged_df_prod = pd.merge(left=product_type, right=product, how='inner', left_on= 'PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE')

eyewear_filter = merged_df_prod.loc[:, "PRODUCT_TYPE_EN"] == "Eyewear"
# merged_df_prod
merged_df_prod.loc[eyewear_filter, ["PRODUCT_NAME"]]

PRODUCT_NAME
15      Polar Sun
16      Polar Ice
17   Polar Sports
18     Polar Wave
19  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [97]:

unieke_addressen = retailer_site.drop_duplicates("ADDRESS1")

s1_merge = pd.merge(left=order_header, right=sales_staff, how='inner', left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE')


s2_merge = pd.merge(left=s1_merge, right=retailer_site, how='inner', left_on='RETAILER_SITE_CODE', right_on='RETAILER_SITE_CODE')


f = s2_merge.loc[:, "POSITION_EN"] == "Branch Manager"


s2_merge.loc[f, ["FIRST_NAME", "LAST_NAME", "RETAILER_SITE_CODE"]].drop_duplicates(["FIRST_NAME", "LAST_NAME"])


FIRST_NAME LAST_NAME RETAILER_SITE_CODE
3574     Bayard     Lopes                335

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [13]:

s3_merge = pd.merge(left=sales_staff, right=order_header, how='left', left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE')
dupes_removed = s3_merge.drop_duplicates(["POSITION_EN", "ORDER_DATE"])
manager_filter = dupes_removed.loc[:, "POSITION_EN"].str.contains("Manager")
dupes_removed.loc[manager_filter, ["POSITION_EN", "ORDER_DATE"]]

POSITION_EN  ORDER_DATE
421     Branch Manager         NaN
422    General Manager         NaN
1276  Regional Manager         NaN
1643  District Manager         NaN
5325    Branch Manager  2022-09-09
5326    Branch Manager  2022-10-31
5327    Branch Manager  2022-12-13
5328    Branch Manager  2022-03-13
5329    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [74]:
s4_merge = pd.merge(left=product, right=product_type, how='left', left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE')
s5_merge = pd.merge(left=s4_merge, right=order_details, how='inner', left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER')

filter = s5_merge['QUANTITY'] > 750

s5_merge.loc[filter, ["PRODUCT_NAME", "PRODUCT_TYPE_EN"]].drop_duplicates(["PRODUCT_NAME", "PRODUCT_TYPE_EN"])

PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [87]:
merged = pd.merge(left=product, right=order_details, how='inner', left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER')

unit_price = merged['UNIT_PRICE'].astype("double")
sale_price = merged['UNIT_SALE_PRICE'].astype("double")

discount = ((unit_price - sale_price) / unit_price) > .4

merged.loc[discount, ["PRODUCT_NAME"]].drop_duplicates("PRODUCT_NAME")

PRODUCT_NAME
2765   Hailstorm Titanium Woods Set
37446             BugShield Natural
37967               BugShield Spray
38306         BugShield Lotion Lite
38772              BugShield Lotion
39561             BugShield Extreme
40799                Sun Shelter 15
41194                Sun Shelter 30

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [104]:
merged = pd.merge(left=order_details, right=returned_item, how='inner', left_on='ORDER_DETAIL_CODE', right_on='ORDER_DETAIL_CODE')
merged2 = pd.merge(left=merged, right=return_reason, how='inner', left_on='RETURN_REASON_CODE', right_on='RETURN_REASON_CODE')


return_quantity = merged2['RETURN_QUANTITY'].astype("double")
quantity = merged2['QUANTITY'].astype("double")

discount = (return_quantity / quantity) > .9

merged2.loc[discount, ["RETURN_DESCRIPTION_EN"]].drop_duplicates("RETURN_DESCRIPTION_EN")

RETURN_DESCRIPTION_EN
0    Wrong product shipped
1   Unsatisfactory product
15   Wrong product ordered

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [129]:
product_en_type =  pd.merge(left=product, right=product_type, how='inner', left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE')
# product_en_type.add(    )
f = product_en_type.groupby("PRODUCT_TYPE_CODE", as_index=False)["PRODUCT_NUMBER"].count().rename(columns= {"PRODUCT_NUMBER": "count"} )
f
# product_en_type.loc[f, :]
# product_en_type

PRODUCT_TYPE_CODE  count
0                  1     10
1                 10      5
2                 11      5
3                 12      5
4                 13      4
5                 14      4
6                 15      5
7                 16      5
8                 17      5
9                 18      4
10                19      4
11                 2      6
12                20      3
13                21      4
14                 3      7
15                 4      6
16                 5     12
17                 6      4
18                 7      4
19                 8      7
20                 9      6

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [131]:
landen =  pd.merge(left=country, right=retailer_site, how='inner', left_on='COUNTRY_CODE', right_on='COUNTRY_CODE')

p = landen.groupby("COUNTRY", as_index=False)["COUNTRY_CODE"].count().rename(columns= {"COUNTRY_CODE": "Hoeveelheid"} )
p

COUNTRY  Hoeveelheid
0        Australia           10
1          Austria           10
2          Belgium           10
3           Brazil            1
4           Canada           31
5            China           10
6          Denmark            6
7          Finland            5
8           France           29
9          Germany           32
10           Italy           12
11           Japan           25
12           Korea           10
13          Mexico           10
14     Netherlands           20
15           Spain            9
16          Sweden           12
17     Switzerland           13
18          Taiwan           10
19  United Kingdom           25
20   United States          101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [109]:
producten =  pd.merge(left=product, right=product_type, how='inner', left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE')
producten =  pd.merge(left=producten, right=order_details, how='inner', left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER')


producten['UNIT_PRICE'] = producten['UNIT_PRICE'].astype('double')

# producten.drop_duplicates()

f = producten["PRODUCT_TYPE_EN"] == "Cooking Gear"
producten = producten.loc[f, :]

producten['Hoeveelheid'] = 0
g = producten.groupby("PRODUCT_NAME", as_index=False)['QUANTITY'].sum()

d = producten.groupby("PRODUCT_NAME", as_index=False)['UNIT_PRICE'].agg(np.mean).round(2)


frame = pd.concat([g, d], axis=True)


frame = frame.loc[:, ~frame.columns.duplicated()].copy()
frame = frame.sort_values(by="QUANTITY", ascending=False)
frame





PRODUCT_NAME  QUANTITY  UNIT_PRICE
9        TrailChef Water Bag     41770        6.18
2              TrailChef Cup     31834        7.14
0          TrailChef Canteen     31712       12.18
6      TrailChef Kitchen Kit     25964       22.74
8         TrailChef Utensils     15928       17.73
4     TrailChef Double Flame     14510      136.60
1         TrailChef Cook Set     12700       52.61
7     TrailChef Single Flame     10614       65.09
5           TrailChef Kettle      9900       12.38
3  TrailChef Deluxe Cook Set      3744      123.54

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2022. (4 kolommen, 4 rijen) 

In [59]:
frame = sales_branch.merge(sales_staff, left_on="SALES_BRANCH_CODE", how="inner", right_on="SALES_BRANCH_CODE")
df2 = frame.merge(target, left_on="SALES_STAFF_CODE", how="inner", right_on="SALES_STAFF_CODE")

filter = (df2['CITY'] == "Amsterdam") & (df2['SALES_YEAR'] == "2022")

df2 = df2.loc[filter, ["FIRST_NAME", "LAST_NAME", "SALES_STAFF_CODE", "SALES_TARGET"]]
df2['SALES_TARGET'] = df2['SALES_TARGET'].astype('int')

g = df2.groupby("SALES_STAFF_CODE", as_index=False)['SALES_TARGET'].agg({'min', 'max'})
g = g.drop(['SALES_STAFF_CODE'], axis='columns')
df2 = df2.drop_duplicates(["FIRST_NAME"]).reset_index(drop=True)


df3 = pd.concat([df2.loc[:, ["FIRST_NAME", "LAST_NAME"]], g], axis=1)
df3

FIRST_NAME         LAST_NAME    max  min
0    Belinda  Jansen-Velasquez  34873   87
1      Ellen           Shapiro  25512   48
2        Jan         Haverkamp  24800   83
3      Maria          Laponder  19825   36

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [19]:
# ? 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [70]:

df = pd.DataFrame(columns=["SALES_STAFF_CODE","FIRST_NAME","LAST_NAME","POSITION_EN","WORK_PHONE","EXTENSION","FAX","EMAIL","DATE_HIRED",	"SALES_BRANCH_CODE", "TRIAL888"])
for index, row in sales_staff.iterrows():
    if order_header['SALES_STAFF_CODE'].str.contains(row['SALES_STAFF_CODE']).any() == False:
        df.loc[len(df.index)] = row
        # row


df.loc[:, ['FIRST_NAME', 'LAST_NAME']]
# len(df)
# len(df)

FIRST_NAME      LAST_NAME
0      Walter         Taylor
1    Penelope        Tamrine
2       Polly         Linton
3       Gracy        Gellens
4        Kick        Kalkman
5       Karin      Bergström
6       Sally          White
7       Frank        Bretton
8     Georges  Saint-Germain
9        Alex      Rodriguez
10   Samantha          Floyd
11      Julie          Olsen
12      Corey         Wright
13    Estelle        Johnson
14       Greg        Belding
15      Julie        Béchard
16      Lucía          Reyna
17     Kazumi        Uragome
18      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [ ]:
# ?

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [36]:
# ?

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114


Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [71]:
# ?

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [73]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [74]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [77]:
frame = sales_staff
frame.rename(columns={"FIRST_NAME": "Medewerker"}, inplace=True)
dienst_data = []

for index, row in frame.iterrows():
    frame["DATE_HIRED"] = pd.to_datetime(frame["DATE_HIRED"], dayfirst=True)

    if date.today().year - date.today().year - row["DATE_HIRED"].year >= 25:
        dienst_data.append("lang in dienst")
    else:
        dienst_data.append("kort in dienst")

frame["jaren_in_dienst"] = dienst_data
frame.loc[:, ["Medewerker", "jaren_in_dienst"]]

Medewerker jaren_in_dienst
0       Tuomas  kort in dienst
1     Chang-ho  kort in dienst
2      Jung-ho  kort in dienst
3    Kwei-Ping  kort in dienst
4    Chin-Tsai  kort in dienst
..         ...             ...
97       Tomás  kort in dienst
98      Bayard  kort in dienst
99      Viviam  kort in dienst
100    Eduardo  kort in dienst
101     Helena  kort in dienst

[102 rows x 2 columns]